src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs

```diff
@@ -1242,92 +1252,105 @@ internal static bool DeepEquals(JsonElement left, JsonElement right)
+                        if (!objectEnumerator2.MoveNext())
                         {
                             return false;
                         }
 
-                        JsonProperty leftProp = leftObjectEnumerator.Current;
-                        JsonProperty rightProp = rightObjectEnumerator.Current;
+                        JsonProperty prop1 = objectEnumerator1.Current;
+                        JsonProperty prop2 = objectEnumerator2.Current;
 
-                        if (!NameEquals(leftProp, rightProp))
+                        if (!NameEquals(prop1, prop2))
                         {
                             // We have our first mismatch, fall back to unordered comparison.
-                            return UnorderedObjectDeepEquals(leftObjectEnumerator, rightObjectEnumerator);
+                            return UnorderedObjectDeepEquals(objectEnumerator1, objectEnumerator2, remainingProps: count);
                         }
 
-                        if (!DeepEquals(leftProp.Value, rightProp.Value))
+                        if (!DeepEquals(prop1.Value, prop2.Value))
                         {
                             return false;
                         }
+
+                        count--;
                     }
 
-                    return !rightObjectEnumerator.MoveNext();
+                    Debug.Assert(!objectEnumerator2.MoveNext());
+                    return true;
 
-                    static bool UnorderedObjectDeepEquals(ObjectEnumerator left, ObjectEnumerator right)
+                    static bool UnorderedObjectDeepEquals(ObjectEnumerator objectEnumerator1, ObjectEnumerator objectEnumerator2, int remainingProps)
                     {
-                        Dictionary<string, JsonElement> rightElements = new(StringComparer.Ordinal);
+                        Dictionary<string, JsonElement> properties2 = new(capacity: remainingProps, StringComparer.Ordinal);
                         do
                         {
-                            JsonProperty prop = right.Current;
-                            rightElements.TryAdd(prop.Name, prop.Value);
+                            JsonProperty prop2 = objectEnumerator2.Current;
+                            properties2.TryAdd(prop2.Name, prop2.Value); // first occurrence wins
                         }
-                        while (right.MoveNext());
+                        while (objectEnumerator2.MoveNext());
 
-                        int leftCount = 0;
                         do
                         {
-                            JsonProperty prop = left.Current;
-                            if (!rightElements.TryGetValue(prop.Name, out JsonElement rightElement) || !DeepEquals(prop.Value, rightElement))
+                            JsonProperty prop = objectEnumerator1.Current;
+                            if (!properties2.TryGetValue(prop.Name, out JsonElement rightElement) || !DeepEquals(prop.Value, rightElement))
```

Comment:
Property name comparison is case-sensitive?

Reply:
JSON properties are always case sensitive, but a serializer could decide to to bind to JSON properties in case insensitive manner.